clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -940,6 +946,134 @@ public void testUnsupportedForMessageFormatInTxnOffsetCommit() {
+            assertEquals(epoch, txnOffsetCommitRequest.data.producerEpoch());
+            return txnOffsetCommitRequest.data.groupInstanceId().equals(instanceId)
+                && !txnOffsetCommitRequest.data.memberId().equals(memberId);
+        }, new TxnOffsetCommitResponse(0, singletonMap(tp, Errors.FENCED_INSTANCE_ID)));
+
+        sender.runOnce();  // TxnOffsetCommit Handled
+
+        assertTrue(transactionManager.hasError());
+        assertTrue(transactionManager.lastError() instanceof CommitFailedException);
+        assertTrue(sendOffsetsResult.isCompleted());
+        assertFalse(sendOffsetsResult.isSuccessful());
+        assertTrue(sendOffsetsResult.error() instanceof CommitFailedException);
+        assertAbortableError(CommitFailedException.class);
+    }
+
+    @Test
+    public void testUnknownMemberIdInTxnOffsetCommitByGroupMetadata() {
+        final String consumerGroupId = "consumer";
+        final long pid = 13131L;
+        final short epoch = 1;
+        final TopicPartition tp = new TopicPartition("foo", 0);
+        final String memberId = "member";
+        final String unknownMemberId = "unknownMember";
+
+        doInitTransactions(pid, epoch);
+
+        transactionManager.beginTransaction();
+
+        TransactionalRequestResult sendOffsetsResult = transactionManager.sendOffsetsToTransaction(
+            singletonMap(tp, new OffsetAndMetadata(39L)),
+            new ConsumerGroupMetadata(consumerGroupId, 5, unknownMemberId, Optional.empty()));
+
+        prepareAddOffsetsToTxnResponse(Errors.NONE, consumerGroupId, pid, epoch);
+        sender.runOnce();  // AddOffsetsToTxn Handled, TxnOffsetCommit Enqueued
+        sender.runOnce();  // FindCoordinator Enqueued
+
+        prepareFindCoordinatorResponse(Errors.NONE, false, CoordinatorType.GROUP, consumerGroupId);
+        sender.runOnce();  // FindCoordinator Returned
+
+        client.prepareResponse(request -> {
+            TxnOffsetCommitRequest txnOffsetCommitRequest = (TxnOffsetCommitRequest) request;
+            assertEquals(consumerGroupId, txnOffsetCommitRequest.data.groupId());
+            assertEquals(pid, txnOffsetCommitRequest.data.producerId());
+            assertEquals(epoch, txnOffsetCommitRequest.data.producerEpoch());
+            return !txnOffsetCommitRequest.data.memberId().equals(memberId);
+        }, new TxnOffsetCommitResponse(0, singletonMap(tp, Errors.UNKNOWN_MEMBER_ID)));
+        sender.runOnce();  // TxnOffsetCommit Handled
+
+        assertTrue(transactionManager.hasError());
+        assertTrue(transactionManager.lastError() instanceof CommitFailedException);
+        assertTrue(sendOffsetsResult.isCompleted());
+        assertFalse(sendOffsetsResult.isSuccessful());
+        assertTrue(sendOffsetsResult.error() instanceof CommitFailedException);
+        assertAbortableError(CommitFailedException.class);
+    }
+
+    @Test
+    public void testIllegalGenerationInTxnOffsetCommitByGroupMetadata() {
```

Comment:
These test cases seem to be identical code other than the error. Can we factor out a helper?

Reply:
Sg